src/ccl_cls.c

```diff
@@ -18,202 +18,271 @@ typedef struct{
+                           ccl_f2d_t *psp,
+                           int nl_out, double *l_out, double *cl_out,
+                           int *status) {
+  int local_status;
+  int clastatus, lind;
+  integ_cl_par ipar;
+  gsl_function F;
+  double lkmin, lkmax, l;
+  double result, eresult;
+  int gslstatus;
+  gsl_integration_workspace *w;
+  gsl_integration_cquad_workspace *w_cquad;
+  size_t nevals;
+
+  // make sure to init core things for safety
+  if (!cosmo->computed_distances) {
+    *status = CCL_ERROR_DISTANCES_INIT;
+    ccl_cosmology_set_status_message(
+      cosmo,
+      "ccl_cls.c: ccl_angular_cl_limber(): distance splines have not been precomputed!");
+    return;
+  }
+
+  // Figure out which power spectrum to use
   ccl_f2d_t *psp_use;
-  if(psp==NULL) {
-    if (!cosmo->computed_power) ccl_cosmology_compute_power(cosmo, status);
-    // Return if computation failed
-    if (!cosmo->computed_power) return NAN;
-    psp_use=cosmo->data.p_nl;
+  if (psp == NULL) {
+    if (!cosmo->computed_nonlin_power) {
+      *status = CCL_ERROR_NONLIN_POWER_INIT;
+      ccl_cosmology_set_status_message(
+        cosmo,
+        "ccl_cls.c: ccl_angular_cl_limber(): non-linear power spctrum has not been computed!");
+      return;
+    }
+    psp_use = cosmo->data.p_nl;
   }
   else
-    psp_use=psp;
+    psp_use = psp;
 
-  int clastatus=0;
-  //Set up integrating function
-  integ_cl_par ipar;
-  gsl_function F;
-  ipar.l=l;
-  ipar.cosmo=cosmo;
-  ipar.trc1=trc1;
-  ipar.trc2=trc2;
-  ipar.psp=psp_use;
-  ipar.status=&clastatus;
-  F.function=&cl_integrand;
-  F.params=&ipar;
-
-  //Get integration limits
-  double lkmin,lkmax;
-  get_k_interval(cosmo,trc1,trc2,l,&lkmin,&lkmax);
-
-  //Integrate
-  double result,eresult;
-  int gslstatus;
-  gsl_integration_workspace *w=gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-  gslstatus=gsl_integration_qag(&F, lkmin, lkmax, 0,
-                                cosmo->gsl_params.INTEGRATION_LIMBER_EPSREL,
-				cosmo->gsl_params.N_ITERATION,
-                                cosmo->gsl_params.INTEGRATION_LIMBER_GAUSS_KRONROD_POINTS,
-                                w, &result, &eresult);
-  gsl_integration_workspace_free(w);
-
-  // Test if a round-off error occured in the evaluation of the integral
-  // If so, try another integration function, more robust but potentially slower
-  if(gslstatus == GSL_EROUND) {
-    ccl_raise_gsl_warning(gslstatus, "ccl_cls.c: ccl_angular_cl_limber(): "
-			  "Default GSL integration failure, attempting backup method.");
-    gsl_integration_cquad_workspace *w_cquad= \
-      gsl_integration_cquad_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-    size_t nevals=0;
-    gslstatus=gsl_integration_cquad(&F, lkmin, lkmax, 0,
-				    cosmo->gsl_params.INTEGRATION_LIMBER_EPSREL,
-				    w_cquad, &result, &eresult, &nevals);
+  #pragma omp parallel private(lind, clastatus, ipar, lkmin, lkmax, result, \
```

Comment:
I usually find it clearer to just declare all private variables inside the parallel `{ }` and never use the `private` clause.
Not asking to modify anything, but rather if you have any thoughts.

Reply:
I don't have strong thoughts either way here. This is how I've always done things so I just continued to do them that way. I did not know that OpenMP was smart enough to understand the intent if you declared vars inside the parallel block!

Reply:
OK, sounds good. It's smart enough as long as you use `default(none)`, which I think is always necessary for safety anyway.